# ABSTRACT

Hyperparameters are parameters that are specified prior to running machine learning algorithms that have a large effect on the predictive power of statistical models. Knowledge of the relative importance of a hyperparameter to an algorithm and its range of values is crucial to hyperparameter tuning and creating effective models. To either experts or non-experts, determining hyperparameters that optimize model performance can be a tedious and difficult task. Therefore, we develop a hyperparameter database that allows users to visualize and understand how to choose hyperparameters that maximize the predictive power of their models. 

The database is created by running millions of hyperparameter values, over thousands of public datasets and calculating the individual conditional expectation of every hyperparameter to the quality of a model.                 

We analyze the **effect of hyperparameters** on algorithms such as                                                  
Distributed Random Forest (DRF),                                                                               
Generalized Linear Model (GLM),                                                                                
Gradient Boosting Machine (GBM),                                                                            
Boosting (XGBoost) and several more.                                                                          
Consequently, the database attempts to provide a one-stop platform for data scientists to identify hyperparameters that have the most effect on their models in order to speed up the process of developing effective predictive models. Moreover, the database will also use these public datasets to build models that can predict hyperparameters without search and for visualizing and teaching concepts such as statistical power and bias/variance tradeoff. The raw data will also be publically available for the research community.


In [1]:
# import h2o package and specific estimator 
import h2o
from h2o.automl import H2OAutoML
import random, os, sys
from datetime import datetime
import pandas as pd
import logging
import csv
import optparse
import time
import json
from distutils.util import strtobool
import psutil

import warnings
warnings.filterwarnings('ignore')

In [2]:
h2o.init(strict_version_check=False) # start h2o

Checking whether there is an H2O instance running at http://localhost:54321 ..... not found.
Attempting to start a local H2O server...
  Java Version: java version "11.0.2" 2019-01-15 LTS; Java(TM) SE Runtime Environment 18.9 (build 11.0.2+9-LTS); Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.2+9-LTS, mixed mode)
  Starting server from /anaconda3/lib/python3.7/site-packages/h2o/backend/bin/h2o.jar
  Ice root: /var/folders/8c/swkczny90t36p2798_2khxfh0000gn/T/tmp7iqnvzxf
  JVM stdout: /var/folders/8c/swkczny90t36p2798_2khxfh0000gn/T/tmp7iqnvzxf/h2o_newzysharma_started_from_python.out
  JVM stderr: /var/folders/8c/swkczny90t36p2798_2khxfh0000gn/T/tmp7iqnvzxf/h2o_newzysharma_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.


--------------------------  ---------------------------------------------------
H2O cluster uptime:         02 secs
H2O cluster timezone:       America/New_York
H2O data parsing timezone:  UTC
H2O cluster version:        3.24.0.1
H2O cluster version age:    13 days
H2O cluster name:           H2O_from_python_newzysharma_wwx0ru
H2O cluster total nodes:    1
H2O cluster free memory:    2 Gb
H2O cluster total cores:    4
H2O cluster allowed cores:  4
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.7.1 final
--------------------------  ---------------------------------------------------

In [3]:
#importing data to the server
hp = h2o.import_file(path="hour.csv")

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [4]:
#Displaying the head
hp.head()

instant dteday season yr mnth hr holiday weekday workingday weathersit temp atemp hum windspeed casual registered cnt 
 
 
 1 2011-01-01 00:00:00 1 0 1 0 0 6 0 1 0.24 0.2879 0.81 0 3 13 16 
 2 2011-01-01 00:00:00 1 0 1 1 0 6 0 1 0.22 0.2727 0.8 0 8 32 40 
 3 2011-01-01 00:00:00 1 0 1 2 0 6 0 1 0.22 0.2727 0.8 0 5 27 32 
 4 2011-01-01 00:00:00 1 0 1 3 0 6 0 1 0.24 0.2879 0.75 0 3 10 13 
 5 2011-01-01 00:00:00 1 0 1 4 0 6 0 1 0.24 0.2879 0.75 0 0 1 1 
 6 2011-01-01 00:00:00 1 0 1 5 0 6 0 2 0.24 0.2576 0.75 0.0896 0 1 1 
 7 2011-01-01 00:00:00 1 0 1 6 0 6 0 1 0.22 0.2727 0.8 0 2 0 2 
 8 2011-01-01 00:00:00 1 0 1 7 0 6 0 1 0.2 0.2576 0.86 0 1 2 3 
 9 2011-01-01 00:00:00 1 0 1 8 0 6 0 1 0.24 0.2879 0.75 0 1 7 8 
 10 2011-01-01 00:00:00 1 0 1 9 0 6 0 1 0.32 0.3485 0.76 0 8 6 14

In [5]:
hp.describe()

Rows:17379
Cols:17




instant dteday season yr mnth hr holiday weekday workingday weathersit temp atemp hum windspeed casual registered cnt 
 
 
 type int time int int int int int int int int real real real real int int int 
 mins 1.0 1293840000000.0 1.0 0.0 1.0 0.0 0.0 0.0 0.0 1.0 0.02 0.0 0.0 0.0 0.0 0.0 1.0 
 mean 8690.0 1325477314552.0461 2.501639910236492 0.5025605615973301 6.5377754761493785 11.546751826917548 0.028770355026181024 3.003682605443351 0.6827205247712756 1.425283387997008 0.4969871684216584 0.47577510213476026 0.6272288394038784 0.1900976063064618 35.67621842453536 153.78686920996606 189.4630876345014 
 maxs 17379.0 1356912000000.0 4.0 1.0 12.0 23.0 1.0 6.0 1.0 4.0 1.0 1.0 1.0 0.8507 367.0 886.0 977.0 
 sigma 5017.029499614288 18150225217.779854 1.1069181394480765 0.5000078290910197 3.438775713750168 6.914405095264493 0.16716527638437123 2.005771456110988 0.4654306335238829 0.6393568777542534 0.19255612124972193 0.17185021563535932 0.19292983406291514 0.1223402285727905 49.30503038705309 151.35728591258314 181.38759909186476 
 zeros 0 0 0 8645 0 726 16879 2502 5514 0 0 2 22 2180 1581 24 0 
 missing 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
 0 1.0 2011-01-01 00:00:00 1.0 0.0 1.0 0.0 0.0 6.0 0.0 1.0 0.24 0.2879 0.81 0.0 3.0 13.0 16.0 
 1 2.0 2011-01-01 00:00:00 1.0 0.0 1.0 1.0 0.0 6.0 0.0 1.0 0.22 0.2727 0.8 0.0 8.0 32.0 40.0 
 2 3.0 2011-01-01 00:00:00 1.0 0.0 1.0 2.0 0.0 6.0 0.0 1.0 0.22 0.2727 0.8 0.0 5.0 27.0 32.0 
 3 4.0 2011-01-01 00:00:00 1.0 0.0 1.0 3.0 0.0 6.0 0.0 1.0 0.24 0.2879 0.75 0.0 3.0 10.0 13.0 
 4 5.0 2011-01-01 00:00:00 1.0 0.0 1.0 4.0 0.0 6.0 0.0 1.0 0.24 0.2879 0.75 0.0 0.0 1.0 1.0 
 5 6.0 2011-01-01 00:00:00 1.0 0.0 1.0 5.0 0.0 6.0 0.0 2.0 0.24 0.2576 0.75 0.0896 0.0 1.0 1.0 
 6 7.0 2011-01-01 00:00:00 1.0 0.0 1.0 6.0 0.0 6.0 0.0 1.0 0.22 0.2727 0.8 0.0 2.0 0.0 2.0 
 7 8.0 2011-01-01 00:00:00 1.0 0.0 1.0 7.0 0.0 6.0 0.0 1.0 0.2 0.2576 0.86 0.0 1.0 2.0 3.0 
 8 9.0 2011-01-01 00:00:00 1.0 0.0 1.0 8.0 0.0 6.0 0.0 1.0 0.24 0.2879 0.75 0.0 1.0 7.0 8.0 
 9 10.0 2011-01-01 00:00:00 1.0 0.0 1.0 9.0 0.0 6.0 0.0 1.0 0.32 0.3485 0.76 0.0 8.0 6.0 14.0

In [29]:
# Functions

def alphabet(n):
  alpha='0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'    
  str=''
  r=len(alpha)-1   
  while len(str)<n:
    i=random.randint(0,r)
    str+=alpha[i]   
  return str
  
  
def set_meta_data(analysis,run_id,server,data,test,model_path,target,run_time,classification,scale,model,balance,balance_threshold,name,path,nthreads,min_mem_size):
  m_data={}
  m_data['start_time'] = time.time()
  m_data['target']=target
  m_data['server_path']=server
  m_data['data_path']=data 
  m_data['test_path']=test
  m_data['max_models']=model
  m_data['run_time']=run_time
  m_data['run_id'] =run_id
  m_data['scale']=scale
  m_data['classification']=classification
  m_data['scale']=False
  m_data['model_path']=model_path
  m_data['balance']=balance
  m_data['balance_threshold']=balance_threshold
  m_data['project'] =name
  m_data['end_time'] = time.time()
  m_data['execution_time'] = 0.0
  m_data['run_path'] =path
  m_data['nthreads'] = nthreads
  m_data['min_mem_size'] = min_mem_size
  m_data['analysis'] = analysis
  return m_data


def dict_to_json(dct,n):
  j = json.dumps(dct, indent=4)
  f = open(n, 'w')
  print(j, file=f)
  f.close()
  
  
def stackedensemble(mod):
    coef_norm=None
    try:
      metalearner = h2o.get_model(mod.metalearner()['name'])
      coef_norm=metalearner.coef_norm()
    except:
      pass        
    return coef_norm

def stackedensemble_df(df):
    bm_algo={ 'GBM': None,'GLM': None,'DRF': None,'XRT': None,'Dee': None}
    for index, row in df.iterrows():
      if len(row['model_id'])>3:
        key=row['model_id'][0:3]
        if key in bm_algo:
          if bm_algo[key] is None:
                bm_algo[key]=row['model_id']
    bm=list(bm_algo.values()) 
    bm=list(filter(None.__ne__, bm))             
    return bm

def se_stats(modl):
    d={}
    d['algo']=modl.algo
    d['model_id']=modl.model_id   
    d['auc']=modl.auc()   
    d['roc']=modl.roc()
    d['mse']=modl.mse()   
    d['null_degrees_of_freedom']=modl.null_degrees_of_freedom()
    d['null_deviance']=modl.null_deviance()
    d['residual_degrees_of_freedom']=modl.residual_degrees_of_freedom()   
    d['residual_deviance']=modl.residual_deviance()
    d['rmse']=modl.rmse()
    return d

def get_model_by_algo(algo,models_dict):
    mod=None
    mod_id=None    
    for m in list(models_dict.keys()):
        if m[0:3]==algo:
            mod_id=m
            mod=h2o.get_model(m)      
    return mod,mod_id     
    
    
def gbm_stats(modl):
    d={}
    d['algo']=modl.algo
    d['model_id']=modl.model_id   
    d['varimp']=modl.varimp()  
    return d
    
    
def dl_stats(modl):
    d={}
    d['algo']=modl.algo
    d['model_id']=modl.model_id   
    d['varimp']=modl.varimp()  
    return d
    
    
def drf_stats(modl):
    d={}
    d['algo']=modl.algo
    d['model_id']=modl.model_id   
    d['varimp']=modl.varimp()  
    d['roc']=modl.roc()      
    return d
    
def xrt_stats(modl):
    d={}
    d['algo']=modl.algo
    d['model_id']=modl.model_id   
    d['varimp']=modl.varimp()  
    d['roc']=modl.roc()      
    return d
    
    
def glm_stats(modl):
    d={}
    d['algo']=modl.algo
    d['model_id']=modl.model_id   
    d['coef']=modl.coef()  
    d['coef_norm']=modl.coef_norm()      
    return d
    
def model_performance_stats(perf):
    d={}
    try:    
      d['mse']=perf.mse()
    except:
      pass      
    try:    
      d['rmse']=perf.rmse() 
    except:
      pass      
    try:    
      d['null_degrees_of_freedom']=perf.null_degrees_of_freedom()
    except:
      pass      
    try:    
      d['residual_degrees_of_freedom']=perf.residual_degrees_of_freedom()
    except:
      pass      
    try:    
      d['residual_deviance']=perf.residual_deviance() 
    except:
      pass      
    try:    
      d['null_deviance']=perf.null_deviance() 
    except:
      pass      
    try:    
      d['aic']=perf.aic() 
    except:
      pass      
    try:
      d['logloss']=perf.logloss() 
    except:
      pass    
    try:
      d['auc']=perf.auc()
    except:
      pass  
    try:
      d['gini']=perf.gini()
    except:
      pass    
    return d
    
def impute_missing_values(df, x, scal=False):
    # determine column types
    ints, reals, enums = [], [], []
    for key, val in df.types.items():
        if key in x:
            if val == 'enum':
                enums.append(key)
            elif val == 'int':
                ints.append(key)            
            else: 
                reals.append(key)    
    _ = df[reals].impute(method='mean')
    _ = df[ints].impute(method='median')
    if scal:
        df[reals] = df[reals].scale()
        df[ints] = df[ints].scale()    
    return


def get_independent_variables(df, targ):
    C = [name for name in df.columns if name != targ]
    # determine column types
    ints, reals, enums = [], [], []
    for key, val in df.types.items():
        if key in C:
            if val == 'enum':
                enums.append(key)
            elif val == 'int':
                ints.append(key)            
            else: 
                reals.append(key)    
    x=ints+enums+reals
    return x
    
def get_all_variables_csv(i):
    ivd={}
    try:
      iv = pd.read_csv(i,header=None)
    except:
      sys.exit(1)    
    col=iv.values.tolist()[0]
    dt=iv.values.tolist()[1]
    i=0
    for c in col:
      ivd[c.strip()]=dt[i].strip()
      i+=1        
    return ivd
    
    

def check_all_variables(df,dct,y=None):     
    targ=list(dct.keys())     
    for key, val in df.types.items():
        if key in targ:
          if dct[key] not in ['real','int','enum']:                      
            targ.remove(key)  
    for key, val in df.types.items():
        if key in targ:            
          if dct[key] != val:
            print('convert ',key,' ',dct[key],' ',val)
            if dct[key]=='enum':
                try:
                  df[key] = df[key].asfactor() 
                except:
                  targ.remove(key)                 
            if dct[key]=='int': 
                try:                
                  df[key] = df[key].asnumeric() 
                except:
                  targ.remove(key)                  
            if dct[key]=='real':
                try:                
                  df[key] = df[key].asnumeric()  
                except:
                  targ.remove(key)                  
    if y is None:
      y=df.columns[-1] 
    if y in targ:
      targ.remove(y)
    else:
      y=targ.pop()            
    return targ    
    
def predictions(mod,data,run_id):
    test = h2o.import_file(data)
    mod_perf=mod_best.model_performance(test)
              
    stats_test={}
    stats_test=model_performance_stats(mod_perf)

    n=run_id+'_test_stats.json'
    dict_to_json(stats_test,n) 

    try:    
      cf=mod_perf.confusion_matrix(metrics=["f1","f2","f0point5","accuracy","precision","recall","specificity","absolute_mcc","min_per_class_accuracy","mean_per_class_accuracy"])
      cf_df=cf[0].table.as_data_frame()
      cf_df.to_csv(run_id+'_test_confusion_matrix.csv')
    except:
      pass

    predictions = mod_best.predict(test)
    predictions_df=test.cbind(predictions).as_data_frame() 
    predictions_df.to_csv(run_id+'_predictions.csv')
    return

def predictions_test(mod,test,run_id):
    mod_perf=mod_best.model_performance(test)          
    stats_test={}
    stats_test=model_performance_stats(mod_perf)
    n=run_id+'_test_stats.json'
    dict_to_json(stats_test,n) 
    try:
      cf=mod_perf.confusion_matrix()
#      cf=mod_perf.confusion_matrix(metrics=["f1","f2","f0point5","accuracy","precision","recall","specificity","absolute_mcc","min_per_class_accuracy","mean_per_class_accuracy"])
      cf_df=cf.table.as_data_frame()
      cf_df.to_csv(run_id+'_test_confusion_matrix.csv')
    except:
      pass
    predictions = mod_best.predict(test)    
    predictions_df=test.cbind(predictions).as_data_frame() 
    predictions_df.to_csv(run_id+'_predictions.csv')
    return predictions

def check_X(x,df):
    for name in x:
        if name not in df.columns:
          x.remove(name)  
    return x    
    
    
def get_stacked_ensemble(lst):
    se=None
    for model in model_set:
      if 'BestOfFamily' in model:
        se=model
    if se is None:     
      for model in model_set:
        if 'AllModels'in model:
          se=model           
    return se       
    
def get_variables_types(df):
    d={}
    for key, val in df.types.items():
        d[key]=val           
    return d    
    
#  End Functions

In [30]:
all_variables=None

In [31]:
data_path='Users/newzysharma/Desktop/Desktop/Machine_Learning/Project'

In [ ]:
run_id=alphabet(9)
if server_path==None:
  server_path=os.path.abspath(os.curdir)
os.chdir(server_path) 
run_dir = os.path.join(server_path,run_id)
os.mkdir(run_dir)
os.chdir(run_dir)    

# run_id to std out
print (run_id)

# Model with 500 seconds

In [6]:
# Assume the following are passed by the user from the web interface

'''
Need a user id and project id?

'''
target='cnt' 
data_file='hour.csv'
run_time=500
run_id='SOME_ID_20180617_221528' # Just some arbitrary ID
server_path='/Users/newzysharma/Desktop/Desktop/Machine_Learning/Project/RunTime_500Secondst'
classification=False
scale=False
max_models=None
balance_y=False # balance_classes=balance_y
balance_threshold=0.2
project ="HyperparameterDB_Project"  # project_name = project

In [7]:
# assign target and inputs for logistic regression
y = target
X = [name for name in hp.columns if name != y]
print(y)
print(X)

cnt
['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered']


In [8]:
# determine column types
ints, reals, enums = [], [], []
for key, val in hp.types.items():
    if key in X:
        if val == 'enum':
            enums.append(key)
        elif val == 'int':
            ints.append(key)            
        else: 
            reals.append(key)

print(ints)
print(enums)
print(reals)

['instant', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'casual', 'registered']
[]
['dteday', 'temp', 'atemp', 'hum', 'windspeed']


In [9]:
# impute missing values
_ = hp[reals].impute(method='mean')
_ = hp[ints].impute(method='median')

if scale:
    hp[reals] = df[reals].scale()
    hp[ints] = df[ints].scale()

In [19]:
# # set target to factor for classification by default or if user specifies classification
# if classification:
#     [y] = hp[y].asfactor()

In [10]:
hp[y].levels()

[]


## Cross-validate rather than take a test training split with 500 seconds

In [10]:
# automl
# runs for run_time seconds then builds a stacked ensemble

aml = H2OAutoML(max_runtime_secs=run_time,project_name = project) # init automl, run for 500 seconds
aml.train(x=X,  
           y=y,
           training_frame=hp)

AutoML progress: |████████████████████████████████████████████████████████| 100%


## Leaderboard

In [11]:
# view leaderboard
lb = aml.leaderboard
lb.head(30)

model_id mean_residual_deviance rmse mse mae rmsle 
 
 
 XGBoost_2_AutoML_20190414_221634 21.6172 4.64943 21.6172 2.69423 0.0711089 
 XGBoost_3_AutoML_20190414_221634 29.4606 5.42777 29.4606 3.60386 0.10642 
 XGBoost_1_AutoML_20190414_221634 31.6527 5.62607 31.6527 3.31933 0.0660312 
 StackedEnsemble_AllModels_AutoML_20190414_221634 124.366 11.1519 124.366 8.30577 0.392323 
 DRF_1_AutoML_20190414_221634 155.346 12.4638 155.346 7.05905 0.0969439 
 StackedEnsemble_BestOfFamily_AutoML_20190414_221634 404.552 20.1135 404.552 15.287 0.565702 
 GLM_grid_1_AutoML_20190414_221634_model_1 32888.6 181.352 32888.6 142.375 1.57006

In [12]:
aml.leader

Model Details
H2OXGBoostEstimator :  XGBoost
Model Key:  XGBoost_2_AutoML_20190414_221634


ModelMetricsRegression: xgboost
** Reported on train data. **

MSE: 4.708110538871864
RMSE: 2.16981808888945
MAE: 1.1708507339125351
RMSLE: 0.029757525600352653
Mean Residual Deviance: 4.708110538871864

ModelMetricsRegression: xgboost
** Reported on cross-validation data. **

MSE: 21.61720309157098
RMSE: 4.649430405068021
MAE: 2.694231865060186
RMSLE: 0.07110888931759722
Mean Residual Deviance: 21.61720309157098
Cross-Validation Metrics Summary: 


mean       sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
----------------------  ---------  -----------  ------------  ------------  ------------  ------------  ------------
mae                     2.6942     0.322615     3.47593       2.78202       2.6041        2.53907       2.06986
mean_residual_deviance  21.6168    3.39657      28.8531       22.6952       22.0155       20.7037       13.8163
mse                     21.6168    3.39657      28.8531       22.6952       22.0155       20.7037       13.8163
r2                      0.999343   0.000102403  0.999124      0.999315      0.999327      0.999372      0.999578
residual_deviance       21.6168    3.39657      28.8531       22.6952       22.0155       20.7037       13.8163
rmse                    4.61894    0.37562      5.37151       4.76394       4.69207       4.55013       3.71703
rmsle                   0.0624597  0.0240318    0.129925      0.0506733     0.0442042     0.0488125     0.0386838

Scoring History: 


timestamp            duration          number_of_trees    training_rmse       training_mae        training_deviance
---  -------------------  ----------------  -----------------  ------------------  ------------------  -------------------
     2019-04-14 22:24:04  2 min 24.285 sec  0.0                261.9286475726579   188.96308763450142  68606.61641924162
     2019-04-14 22:24:05  2 min 24.503 sec  5.0                204.7985263826639   146.8314604541568   41942.436408510686
     2019-04-14 22:24:05  2 min 24.680 sec  10.0               160.1791867447757   114.00802879522404  25657.371866217727
     2019-04-14 22:24:05  2 min 24.882 sec  15.0               125.03038707826089  88.50036830502266   15632.597692939748
     2019-04-14 22:24:05  2 min 25.053 sec  20.0               97.31989575362786   68.64214397179833   9471.162109496994
---  ---                  ---               ---                ---                 ---                 ---
     2019-04-14 22:24:28  2 min 48.392 sec  195.0              2.350672621033033   1.2583852130213393  5.525661771274309
     2019-04-14 22:24:30  2 min 49.640 sec  200.0              2.2892984622352954  1.2295200778762507  5.240887449192889
     2019-04-14 22:24:31  2 min 50.728 sec  205.0              2.237118601243747   1.2034828715055101  5.004699636030779
     2019-04-14 22:24:32  2 min 51.849 sec  210.0              2.1833596973733633  1.1766797696681872  4.767059568114305
     2019-04-14 22:24:33  2 min 52.811 sec  211.0              2.16981808888945    1.1708507339125351  4.708110538871864


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
registered  1.93911e+09            1                    0.556715
casual      7.15683e+08            0.369078             0.205471
hr          5.07491e+08            0.261713             0.1457
workingday  8.98284e+07            0.0463246            0.0257896
instant     6.94033e+07            0.0357913            0.0199256
atemp       5.75791e+07            0.0296936            0.0165308
temp        4.74625e+07            0.0244764            0.0136264
dteday      3.14025e+07            0.0161943            0.0090156
weekday     1.58227e+07            0.00815976           0.00454266
weathersit  3.8962e+06             0.00200927           0.00111859
hum         3.46073e+06            0.0017847            0.000993569
windspeed   562857                 0.000290266          0.000161595
season      539699                 0.000278323          0.000154947
mnth        410562                 0.000211727          0.000117872
holiday     341838                 0.000176286          9.81411e-05
yr          133964                 6.90855e-05          3.84609e-05

In [13]:
aml.leader.algo

'xgboost'

## Ensemble Exploration

In [14]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
aml_leaderboard_df

model_id  mean_residual_deviance  \
0                   XGBoost_2_AutoML_20190414_221634               21.617203   
1                   XGBoost_3_AutoML_20190414_221634               29.460637   
2                   XGBoost_1_AutoML_20190414_221634               31.652709   
3   StackedEnsemble_AllModels_AutoML_20190414_221634              124.365918   
4                       DRF_1_AutoML_20190414_221634              155.346367   
5  StackedEnsemble_BestOfFamily_AutoML_20190414_2...              404.552066   
6          GLM_grid_1_AutoML_20190414_221634_model_1            32888.626822   

         rmse           mse         mae     rmsle  
0    4.649430     21.617203    2.694232  0.071109  
1    5.427765     29.460637    3.603859  0.106420  
2    5.626074     31.652709    3.319334  0.066031  
3   11.151947    124.365918    8.305767  0.392323  
4   12.463802    155.346367    7.059050  0.096944  
5   20.113480    404.552066   15.286993  0.565702  
6  181.352218  32888.626822  142.374821  1.570058

## Getting Models

### Parameters for XGBoost_1_AutoML_20190408_134347

In [15]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[0])

In [16]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'XGBoost_2_AutoML_20190414_221634',
   'type': 'Key<Model>',
   'URL': '/3/Models/XGBoost_2_AutoML_20190414_221634'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'fold_assignment': {'default': 'AUTO', 'actual': 'Modulo'},
 'fold_column': {'defau

In [46]:
hy_parameter_xgboost = mod_best.params
n=run_id+'hyperparameets_XGBoost_1_AutoML_20190408_134347.json'
dict_to_json(hy_parameter,n)

In [35]:
# Get stacked ensemble  
se=get_stacked_ensemble(model_set)

In [36]:
print(se)

StackedEnsemble_BestOfFamily_AutoML_20190414_221634


In [37]:
if se is not None:
  mod_best=h2o.get_model(se)

In [38]:
mod_best.actual_params

In [39]:
mod_best._id

'StackedEnsemble_BestOfFamily_AutoML_20190414_221634'

In [40]:
mod_best._get_metrics

<function h2o.model.model_base.ModelBase._get_metrics(o, train, valid, xval)>

In [41]:
type(mod_best)

h2o.estimators.stackedensemble.H2OStackedEnsembleEstimator

In [42]:
mods=mod_best.coef_norm
print(mods)

Model Details
H2OStackedEnsembleEstimator :  Stacked Ensemble
Model Key:  StackedEnsemble_BestOfFamily_AutoML_20190414_221634
No model summary for this model


ModelMetricsRegressionGLM: stackedensemble
** Reported on train data. **

MSE: 228.05154670500258
RMSE: 15.101375656045464
MAE: 11.63634317115643
RMSLE: 0.49716685876783684
R^2: 0.993068251012167
Mean Residual Deviance: 228.05154670500258
Null degrees of freedom: 17378
Residual degrees of freedom: 17375
Null deviance: 571761591.0707175
Residual deviance: 3963307.83018624
AIC: 143689.9918584112

ModelMetricsRegressionGLM: stackedensemble
** Reported on cross-validation data. **

MSE: 404.55206573805515
RMSE: 20.113479702380072
MAE: 15.286993185084869
RMSLE: 0.5657017501133251
R^2: 0.9877034231395371
Mean Residual Deviance: 404.55206573805515
Null degrees of freedom: 17378
Residual degrees of freedom: 17375
Null deviance: 571800073.8579907
Residual deviance: 7030710.35046166
AIC: 153651.78688213223
<bound method ModelBase.coef_nor

In [44]:
#%matplotlib inline
#mod_best.std_coef_plot()

### Parameters for StackedEnsemble_AllModels_AutoML_20190408_134347

In [47]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[4])

In [48]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'DRF_1_AutoML_20190414_221634',
   'type': 'Key<Model>',
   'URL': '/3/Models/DRF_1_AutoML_20190414_221634'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'score_tree_interval': {'default': 0, 'actual': 0},
 'fold_assignment': {'default': 'AUTO',

In [49]:
hy_parameter_stack = mod_best.params
n=run_id+'StackedEnsemble_AllModels_AutoML_20190408_171007.json'
dict_to_json(hy_parameter,n)

### BUG:  I am trying to get Parameters for model 6(GBM_1_AutoML_20190408_134347) but i am getting some other model but i am getting GBM_4_AutoML_20190408_171007

In [105]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[6])

In [106]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'GBM_4_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/GBM_4_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'score_tree_interval': {'default': 0, 'actual': 5},
 'fold_assignment': {'default': 'AUTO',

# Model with runtime 1000 seconds

In [57]:
# Assume the following are passed by the user from the web interface

'''
Need a user id and project id?

'''
target='cnt' 
data_file='hour.csv'
run_time=1000
run_id='SOME_ID_20180617_221529' # Just some arbitrary ID
server_path='Users/newzysharma/Desktop/Desktop/Machine_Learning/Project'
classification=False
scale=False
max_models=None
balance_y=False # balance_classes=balance_y
balance_threshold=0.2
project ="HyperparameterDB_Project"  # project_name = project

In [58]:
# Use local data file or download from some type of bucket
import os

data_path=os.path.join(server_path,data_file)
data_path

'Users/newzysharma/Desktop/Desktop/Machine_Learning/Project/hour.csv'

In [59]:
# assign target and inputs for logistic regression
y = target
X = [name for name in hp.columns if name != y]
print(y)
print(X)

cnt
['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered']


In [60]:
# determine column types
ints, reals, enums = [], [], []
for key, val in hp.types.items():
    if key in X:
        if val == 'enum':
            enums.append(key)
        elif val == 'int':
            ints.append(key)            
        else: 
            reals.append(key)

print(ints)
print(enums)
print(reals)

['instant', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'casual', 'registered']
[]
['dteday', 'temp', 'atemp', 'hum', 'windspeed']


In [61]:
# impute missing values
_ = hp[reals].impute(method='mean')
_ = hp[ints].impute(method='median')

if scale:
    hp[reals] = hp[reals].scale()
    hp[ints] = hp[ints].scale()

In [20]:
# set target to factor for classification by default or if user specifies classification
#if classification:
    #hp[y] = hp[y].asfactor()

In [31]:
hp[y].levels()

[]

In [62]:
if classification:
    class_percentage = y_balance=df[y].mean()[0]/(df[y].max()-df[y].min())
    if class_percentage < balance_threshold:
        balance_y=True
        

print(run_time)
type(run_time)

1000


int


## Cross-validate rather than take a test training split with 1000 seconds

In [63]:
# automl
# runs for run_time seconds then builds a stacked ensemble

aml = H2OAutoML(max_runtime_secs=run_time,project_name = project) # init automl, run for 1000 seconds
aml.train(x=X,  
           y=y,
           training_frame=hp)

AutoML progress: |████████████████████████████████████████████████████████| 100%


## Leaderboard

In [64]:
# view leaderboard
lb = aml.leaderboard
lb

model_id mean_residual_deviance rmse mse mae rmsle 
 
 
 GBM_2_AutoML_20190408_171007 10.3989 3.22474 10.3989 1.88964 0.0353651 
 GBM_1_AutoML_20190408_171007 13.2483 3.63982 13.2483 2.30598 0.0573338 
 XGBoost_1_AutoML_20190408_171007 14.936 3.86471 14.936 2.25298 0.0426858 
 GBM_3_AutoML_20190408_171007 22.3503 4.72761 22.3503 2.93989 0.0688736 
 StackedEnsemble_AllModels_AutoML_20190408_171007 23.1046 4.80673 23.1046 3.22004 0.180671 
 XGBoost_3_AutoML_20190408_171007 23.4324 4.84071 23.4324 3.16966 0.0785482 
 GBM_4_AutoML_20190408_171007 25.8815 5.08739 25.8815 2.98719 nan 
 XGBoost_2_AutoML_20190408_171007 31.3904 5.60271 31.3904 3.26648 0.0692559 
 GLM_grid_1_AutoML_20190408_171007_model_1 38.387 6.19573 38.387 4.45197 nan 
 DRF_1_AutoML_20190408_171007 44.9979 6.70804 44.9979 3.64084 0.0762501

In [65]:
aml.leader

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  GBM_2_AutoML_20190408_171007


ModelMetricsRegression: gbm
** Reported on train data. **

MSE: 2.7545139005515074
RMSE: 1.659672829370749
MAE: 1.1420248950973013
RMSLE: 0.026260655890205944
Mean Residual Deviance: 2.7545139005515074

ModelMetricsRegression: gbm
** Reported on cross-validation data. **

MSE: 10.398929973581447
RMSE: 3.2247371944984056
MAE: 1.8896396840192615
RMSLE: 0.03536506290104928
Mean Residual Deviance: 10.398929973581447
Cross-Validation Metrics Summary: 


mean       sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
----------------------  ---------  -----------  ------------  ------------  ------------  ------------  ------------
mae                     1.88964    0.0331531    1.88882       1.84409       1.95555       1.83307       1.92668
mean_residual_deviance  10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
mse                     10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
r2                      0.999684   3.44699e-05  0.999712      0.999737      0.999593      0.999687      0.999689
residual_deviance       10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
rmse                    3.21618    0.166        3.07971       2.95054       3.64714       3.21238       3.19112
rmsle                   0.0353167  0.00130899   0.0337463     0.0375869     0.0355834     0.0327149     0.0369519

Scoring History: 


timestamp            duration    number_of_trees    training_rmse       training_mae        training_deviance
---  -------------------  ----------  -----------------  ------------------  ------------------  -------------------
     2019-04-08 17:25:27  16.816 sec  0.0                181.38238043116908  142.3998489274107   32899.56793087735
     2019-04-08 17:25:27  16.882 sec  5.0                107.42392635466818  84.24003602830221   11539.899953453174
     2019-04-08 17:25:27  16.935 sec  10.0               63.894083815772795  50.023308890770906  4082.453946656999
     2019-04-08 17:25:27  16.985 sec  15.0               38.450969966077004  30.034970414165908  1478.4770913321556
     2019-04-08 17:25:27  17.034 sec  20.0               25.653038601608305  19.734860646038076  658.0783894956057
---  ---                  ---         ---                ---                 ---                 ---
     2019-04-08 17:25:30  19.344 sec  270.0              1.7050142079828006  1.1700358492018363  2.9070734494232164
     2019-04-08 17:25:30  19.393 sec  275.0              1.6909674134211323  1.1622558060178851  2.859370793252155
     2019-04-08 17:25:30  19.451 sec  280.0              1.677069230591134   1.1544659421841645  2.812561204195539
     2019-04-08 17:25:30  19.520 sec  285.0              1.66171661485495    1.1441283714813113  2.761302108084994
     2019-04-08 17:25:30  19.534 sec  286.0              1.659672829370749   1.1420248950973013  2.7545139005515074


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
registered  2.56858e+09            1                    0.918241
casual      1.5348e+08             0.0597529            0.0548675
hr          4.37214e+07            0.0170216            0.0156299
instant     9.33502e+06            0.00363431           0.00333717
workingday  7.97255e+06            0.00310387           0.0028501
dteday      5.19893e+06            0.00202404           0.00185856
temp        3.35408e+06            0.00130581           0.00119905
weekday     2.51315e+06            0.000978418          0.000898423
atemp       2.09453e+06            0.000815442          0.000748772
hum         538359                 0.000209594          0.000192457
season      207410                 8.07489e-05          7.41469e-05
weathersit  139702                 5.43889e-05          4.99421e-05
mnth        64992.2                2.53027e-05          2.3234e-05
windspeed   55394.5                2.15662e-05          1.98029e-05
yr          19290.6                7.51022e-06          6.89619e-06
holiday     9624.22                3.7469e-06           3.44055e-06

In [66]:
aml.leader.algo

'gbm'

## Ensemble Exploration

In [67]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
aml_leaderboard_df

model_id  mean_residual_deviance  \
0                        GBM_2_AutoML_20190408_171007               10.398930   
1                        GBM_1_AutoML_20190408_171007               13.248308   
2                    XGBoost_1_AutoML_20190408_171007               14.936004   
3                        GBM_3_AutoML_20190408_171007               22.350299   
4    StackedEnsemble_AllModels_AutoML_20190408_171007               23.104649   
5                    XGBoost_3_AutoML_20190408_171007               23.432444   
6                        GBM_4_AutoML_20190408_171007               25.881513   
7                    XGBoost_2_AutoML_20190408_171007               31.390413   
8           GLM_grid_1_AutoML_20190408_171007_model_1               38.387049   
9                        DRF_1_AutoML_20190408_171007               44.997852   
10              DeepLearning_1_AutoML_20190408_171007               51.970420   
11  StackedEnsemble_BestOfFamily_AutoML_20190408_1...               58.092277   
12                       GBM_5_AutoML_20190408_171007               82.943169   
13                       XRT_1_AutoML_20190408_171007              156.118172   

         rmse         mse       mae     rmsle  
0    3.224737   10.398930  1.889640  0.035365  
1    3.639823   13.248308  2.305975  0.057334  
2    3.864713   14.936004  2.252979  0.042686  
3    4.727610   22.350299  2.939890  0.068874  
4    4.806730   23.104649  3.220036  0.180671  
5    4.840707   23.432444  3.169660  0.078548  
6    5.087388   25.881513  2.987189       NaN  
7    5.602715   31.390413  3.266475  0.069256  
8    6.195728   38.387049  4.451969       NaN  
9    6.708044   44.997852  3.640842  0.076250  
10   7.209051   51.970420  4.818606       NaN  
11   7.621829   58.092277  5.466349  0.290238  
12   9.107314   82.943169  5.244854       NaN  
13  12.494726  156.118172  6.946500  0.099340

##### As per the rmse metric that "The smaller the RMSE value, the better the model".
##### As per the mse metric that "The smaller the MSE value, the better the model".
So, In our case GBM_2_AutoML_20190408_171007 has the smallest rmse = 3.224737 and mse = 10.398930 that is why it is our best model.

## Getting Models

### Parameters for GBM_2_AutoML_20190408_171007

In [69]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[0])

In [70]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'GBM_2_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/GBM_2_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'score_tree_interval': {'default': 0, 'actual': 5},
 'fold_assignment': {'default': 'AUTO',

### Parameters for XGBoost_1_AutoML_20190408_171007

In [82]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[2])

In [83]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'XGBoost_1_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/XGBoost_1_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'fold_assignment': {'default': 'AUTO', 'actual': 'Modulo'},
 'fold_column': {'defau

### Parameters for StackedEnsemble_AllModels_AutoML_20190408_171007

In [86]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[4])

In [87]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'StackedEnsemble_AllModels_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/StackedEnsemble_AllModels_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'response_column': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ColSpecifierV3',
    'schema_type': 'VecSpecifier'},
   'column_name': 'cnt',
   'is_member_of_frames': None}},
 'validation_frame': {'default': None, 'actual': None},
 'blending_frame': {'default': None, 'actual': None},
 'base_models': {'default': [],
  'actual': [{'__meta': {'schema_version': 3,
     'schema_name': 'ModelKey

## Model with runtime 1350 seconds

In [128]:
# Assume the following are passed by the user from the web interface

'''
Need a user id and project id?

'''
target='cnt' 
data_file='hour.csv'
run_time=1350
run_id='SOME_ID_20180617_221530' # Just some arbitrary ID
server_path='Users/newzysharma/Desktop/Desktop/Machine_Learning/Project'
classification=False
scale=False
max_models=None
balance_y=False # balance_classes=balance_y
balance_threshold=0.2
project ="HyperparameterDB_Project"  # project_name = project

In [109]:
# Use local data file or download from some type of bucket
import os

data_path=os.path.join(server_path,data_file)
data_path

'Users/newzysharma/Desktop/Desktop/Machine_Learning/Project/hour.csv'

In [110]:
# assign target and inputs for logistic regression
y = target
X = [name for name in hp.columns if name != y]
print(y)
print(X)

cnt
['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered']


In [111]:
# determine column types
ints, reals, enums = [], [], []
for key, val in hp.types.items():
    if key in X:
        if val == 'enum':
            enums.append(key)
        elif val == 'int':
            ints.append(key)            
        else: 
            reals.append(key)

print(ints)
print(enums)
print(reals)

['instant', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'casual', 'registered']
[]
['dteday', 'temp', 'atemp', 'hum', 'windspeed']


In [112]:
# impute missing values
_ = hp[reals].impute(method='mean')
_ = hp[ints].impute(method='median')

if scale:
    hp[reals] = hp[reals].scale()
    hp[ints] = hp[ints].scale()

In [113]:
if classification:
    class_percentage = y_balance=df[y].mean()[0]/(df[y].max()-df[y].min())
    if class_percentage < balance_threshold:
        balance_y=True
        

print(run_time)
type(run_time)

1350


int

## Cross-validate rather than take a test training split with 1350 seconds

In [114]:
# automl
# runs for run_time seconds then builds a stacked ensemble

aml = H2OAutoML(max_runtime_secs=run_time,project_name = project) # init automl, run for 1350 seconds
aml.train(x=X,  
           y=y,
           training_frame=hp)

AutoML progress: |████████████████████████████████████████████████████████| 100%


## Leaderboard

In [115]:
# view leaderboard
lb = aml.leaderboard
lb

model_id mean_residual_deviance rmse mse mae rmsle 
 
 
 GBM_2_AutoML_20190408_171007 10.3989 3.22474 10.3989 1.88964 0.0353651 
 GBM_1_AutoML_20190408_171007 13.2483 3.63982 13.2483 2.30598 0.0573338 
 GBM_4_AutoML_20190408_190707 14.2118 3.76985 14.2118 2.15209 0.0367956 
 XGBoost_1_AutoML_20190408_171007 14.936 3.86471 14.936 2.25298 0.0426858 
 GBM_3_AutoML_20190408_190707 15.2039 3.89922 15.2039 2.27726 0.0438891 
 GBM_1_AutoML_20190408_190707 19.1514 4.37623 19.1514 2.70842 nan 
 XGBoost_2_AutoML_20190408_190707 19.5511 4.42167 19.5511 2.52855 0.050435 
 XGBoost_1_AutoML_20190408_190707 20.9605 4.57826 20.9605 2.65315 0.0493793 
 GBM_2_AutoML_20190408_190707 22.3321 4.72569 22.3321 2.82382 nan 
 GBM_3_AutoML_20190408_171007 22.3503 4.72761 22.3503 2.93989 0.0688736

In [116]:
aml.leader

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  GBM_2_AutoML_20190408_171007


ModelMetricsRegression: gbm
** Reported on train data. **

MSE: 2.7545139005515074
RMSE: 1.659672829370749
MAE: 1.1420248950973013
RMSLE: 0.026260655890205944
Mean Residual Deviance: 2.7545139005515074

ModelMetricsRegression: gbm
** Reported on cross-validation data. **

MSE: 10.398929973581447
RMSE: 3.2247371944984056
MAE: 1.8896396840192615
RMSLE: 0.03536506290104928
Mean Residual Deviance: 10.398929973581447
Cross-Validation Metrics Summary: 


mean       sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
----------------------  ---------  -----------  ------------  ------------  ------------  ------------  ------------
mae                     1.88964    0.0331531    1.88882       1.84409       1.95555       1.83307       1.92668
mean_residual_deviance  10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
mse                     10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
r2                      0.999684   3.44699e-05  0.999712      0.999737      0.999593      0.999687      0.999689
residual_deviance       10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
rmse                    3.21618    0.166        3.07971       2.95054       3.64714       3.21238       3.19112
rmsle                   0.0353167  0.00130899   0.0337463     0.0375869     0.0355834     0.0327149     0.0369519

Scoring History: 


timestamp            duration    number_of_trees    training_rmse       training_mae        training_deviance
---  -------------------  ----------  -----------------  ------------------  ------------------  -------------------
     2019-04-08 17:25:27  16.816 sec  0.0                181.38238043116908  142.3998489274107   32899.56793087735
     2019-04-08 17:25:27  16.882 sec  5.0                107.42392635466818  84.24003602830221   11539.899953453174
     2019-04-08 17:25:27  16.935 sec  10.0               63.894083815772795  50.023308890770906  4082.453946656999
     2019-04-08 17:25:27  16.985 sec  15.0               38.450969966077004  30.034970414165908  1478.4770913321556
     2019-04-08 17:25:27  17.034 sec  20.0               25.653038601608305  19.734860646038076  658.0783894956057
---  ---                  ---         ---                ---                 ---                 ---
     2019-04-08 17:25:30  19.344 sec  270.0              1.7050142079828006  1.1700358492018363  2.9070734494232164
     2019-04-08 17:25:30  19.393 sec  275.0              1.6909674134211323  1.1622558060178851  2.859370793252155
     2019-04-08 17:25:30  19.451 sec  280.0              1.677069230591134   1.1544659421841645  2.812561204195539
     2019-04-08 17:25:30  19.520 sec  285.0              1.66171661485495    1.1441283714813113  2.761302108084994
     2019-04-08 17:25:30  19.534 sec  286.0              1.659672829370749   1.1420248950973013  2.7545139005515074


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
registered  2.56858e+09            1                    0.918241
casual      1.5348e+08             0.0597529            0.0548675
hr          4.37214e+07            0.0170216            0.0156299
instant     9.33502e+06            0.00363431           0.00333717
workingday  7.97255e+06            0.00310387           0.0028501
dteday      5.19893e+06            0.00202404           0.00185856
temp        3.35408e+06            0.00130581           0.00119905
weekday     2.51315e+06            0.000978418          0.000898423
atemp       2.09453e+06            0.000815442          0.000748772
hum         538359                 0.000209594          0.000192457
season      207410                 8.07489e-05          7.41469e-05
weathersit  139702                 5.43889e-05          4.99421e-05
mnth        64992.2                2.53027e-05          2.3234e-05
windspeed   55394.5                2.15662e-05          1.98029e-05
yr          19290.6                7.51022e-06          6.89619e-06
holiday     9624.22                3.7469e-06           3.44055e-06

In [118]:
aml.leader.algo

'gbm'

## Ensemble Exploration

In [119]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
aml_leaderboard_df

model_id  mean_residual_deviance  \
0                        GBM_2_AutoML_20190408_171007               10.398930   
1                        GBM_1_AutoML_20190408_171007               13.248308   
2                        GBM_4_AutoML_20190408_190707               14.211758   
3                    XGBoost_1_AutoML_20190408_171007               14.936004   
4                        GBM_3_AutoML_20190408_190707               15.203899   
5                        GBM_1_AutoML_20190408_190707               19.151410   
6                    XGBoost_2_AutoML_20190408_190707               19.551146   
7                    XGBoost_1_AutoML_20190408_190707               20.960508   
8                        GBM_2_AutoML_20190408_190707               22.332141   
9                        GBM_3_AutoML_20190408_171007               22.350299   
10   StackedEnsemble_AllModels_AutoML_20190408_171007               23.104649   
11                   XGBoost_3_AutoML_20190408_171007               23.432444   
12                       GBM_4_AutoML_20190408_171007               25.881513   
13                   XGBoost_3_AutoML_20190408_190707               28.958293   
14              DeepLearning_1_AutoML_20190408_190707               31.081973   
15                   XGBoost_2_AutoML_20190408_171007               31.390413   
16          GLM_grid_1_AutoML_20190408_190707_model_1               38.387049   
17          GLM_grid_1_AutoML_20190408_171007_model_1               38.387049   
18                       XRT_1_AutoML_20190408_190707               43.280700   
19      XGBoost_grid_1_AutoML_20190408_190707_model_1               44.578618   
20                       DRF_1_AutoML_20190408_171007               44.997852   
21              DeepLearning_1_AutoML_20190408_171007               51.970420   
22                       DRF_1_AutoML_20190408_190707               55.530565   
23  StackedEnsemble_BestOfFamily_AutoML_20190408_1...               58.092277   
24                       GBM_5_AutoML_20190408_171007               82.943169   
25                       GBM_5_AutoML_20190408_190707               83.974272   
26                       XRT_1_AutoML_20190408_171007              156.118172   
27  DeepLearning_grid_1_AutoML_20190408_190707_mod...             1009.056343   
28  DeepLearning_grid_1_AutoML_20190408_190707_mod...             1703.482385   
29      XGBoost_grid_1_AutoML_20190408_190707_model_2             2898.053915   
30          GBM_grid_1_AutoML_20190408_190707_model_1             2958.037238   
31          GBM_grid_1_AutoML_20190408_190707_model_2            16178.386344   
32          GBM_grid_1_AutoML_20190408_190707_model_3            30409.503537   
33          GBM_grid_1_AutoML_20190408_190707_model_4            32572.387845   

          rmse           mse         mae     rmsle  
0     3.224737     10.398930    1.889640  0.035365  
1     3.639823     13.248308    2.305975  0.057334  
2     3.769849     14.211758    2.152092  0.036796  
3     3.864713     14.936004    2.252979  0.042686  
4     3.899218     15.203899    2.277258  0.043889  
5     4.376232     19.151410    2.708416       NaN  
6     4.421668     19.551146    2.528554  0.050435  
7     4.578265     20.960508    2.653145  0.049379  
8     4.725689     22.332141    2.823816       NaN  
9     4.727610     22.350299    2.939890  0.068874  
10    4.806730     23.104649    3.220036  0.180671  
11    4.840707     23.432444    3.169660  0.078548  
12    5.087388     25.881513    2.987189       NaN  
13    5.381291     28.958293    3.536292  0.103629  
14    5.575121     31.081973    3.294086       NaN  
15    5.602715     31.390413    3.266475  0.069256  
16    6.195728     38.387049    4.451969       NaN  
17    6.195728     38.387049    4.451969       NaN  
18    6.578807     43.280700    3.553485  0.076789  
19    6.676722     44.578618    3.718802  0.037516  
20    6.708044     44.997852    3.640842  0.076250  
21    7.209051     51.970420    4.818606   

## Getting models

### Parameters for GBM_2_AutoML_20190408_171007

In [121]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[0])

In [122]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'GBM_2_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/GBM_2_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'score_tree_interval': {'default': 0, 'actual': 5},
 'fold_assignment': {'default': 'AUTO',

### Parameters for XGBoost_1_AutoML_20190408_171007

In [124]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[3])

In [125]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'XGBoost_1_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/XGBoost_1_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'fold_assignment': {'default': 'AUTO', 'actual': 'Modulo'},
 'fold_column': {'defau

### Parameters for DRF_1_AutoML_20190408_171007

In [126]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[20])

In [127]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'DRF_1_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/DRF_1_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'score_tree_interval': {'default': 0, 'actual': 0},
 'fold_assignment': {'default': 'AUTO',

## Model with runtime 1500 seconds

In [134]:
# Assume the following are passed by the user from the web interface

'''
Need a user id and project id?

'''
target='cnt' 
data_file='hour.csv'
run_time=1500
run_id='SOME_ID_20180617_221531' # Just some arbitrary ID
server_path='Users/newzysharma/Desktop/Desktop/Machine_Learning/Project'
classification=False
scale=False
max_models=None
balance_y=False # balance_classes=balance_y
balance_threshold=0.2
project ="HyperparameterDB_Project"  # project_name = project

In [135]:
# Use local data file or download from some type of bucket
import os

data_path=os.path.join(server_path,data_file)
data_path

'Users/newzysharma/Desktop/Desktop/Machine_Learning/Project/hour.csv'

In [136]:
# assign target and inputs for logistic regression
y = target
X = [name for name in hp.columns if name != y]
print(y)
print(X)

cnt
['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered']


In [137]:
# impute missing values
_ = hp[reals].impute(method='mean')
_ = hp[ints].impute(method='median')

if scale:
    hp[reals] = hp[reals].scale()
    hp[ints] = hp[ints].scale()

In [138]:
if classification:
    class_percentage = y_balance=df[y].mean()[0]/(df[y].max()-df[y].min())
    if class_percentage < balance_threshold:
        balance_y=True
        

print(run_time)
type(run_time)

1500


int

## Cross-validate rather than take a test training split with 1500 seconds

In [139]:
# automl
# runs for run_time seconds then builds a stacked ensemble

aml = H2OAutoML(max_runtime_secs=run_time,project_name = project) # init automl, run for 1500 seconds
aml.train(x=X,  
           y=y,
           training_frame=hp)

AutoML progress: |████████████████████████████████████████████████████████| 100%


## Leaderboard

In [140]:
# view leaderboard
lb = aml.leaderboard
lb

model_id mean_residual_deviance rmse mse mae rmsle 
 
 
 GBM_2_AutoML_20190408_171007 10.3989 3.22474 10.3989 1.88964 0.0353651 
 GBM_2_AutoML_20190408_194251 12.7928 3.57671 12.7928 2.13437 0.0450689 
 GBM_1_AutoML_20190408_171007 13.2483 3.63982 13.2483 2.30598 0.0573338 
 GBM_4_AutoML_20190408_190707 14.2118 3.76985 14.2118 2.15209 0.0367956 
 GBM_3_AutoML_20190408_194251 14.2999 3.78152 14.2999 2.17271 0.0358335 
 XGBoost_1_AutoML_20190408_171007 14.936 3.86471 14.936 2.25298 0.0426858 
 GBM_3_AutoML_20190408_190707 15.2039 3.89922 15.2039 2.27726 0.0438891 
 XGBoost_3_AutoML_20190408_194251 15.9838 3.99797 15.9838 2.48441 nan 
 GBM_1_AutoML_20190408_190707 19.1514 4.37623 19.1514 2.70842 nan 
 XGBoost_2_AutoML_20190408_190707 19.5511 4.42167 19.5511 2.52855 0.050435

In [141]:
aml.leader

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  GBM_2_AutoML_20190408_171007


ModelMetricsRegression: gbm
** Reported on train data. **

MSE: 2.7545139005515074
RMSE: 1.659672829370749
MAE: 1.1420248950973013
RMSLE: 0.026260655890205944
Mean Residual Deviance: 2.7545139005515074

ModelMetricsRegression: gbm
** Reported on cross-validation data. **

MSE: 10.398929973581447
RMSE: 3.2247371944984056
MAE: 1.8896396840192615
RMSLE: 0.03536506290104928
Mean Residual Deviance: 10.398929973581447
Cross-Validation Metrics Summary: 


mean       sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
----------------------  ---------  -----------  ------------  ------------  ------------  ------------  ------------
mae                     1.88964    0.0331531    1.88882       1.84409       1.95555       1.83307       1.92668
mean_residual_deviance  10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
mse                     10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
r2                      0.999684   3.44699e-05  0.999712      0.999737      0.999593      0.999687      0.999689
residual_deviance       10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
rmse                    3.21618    0.166        3.07971       2.95054       3.64714       3.21238       3.19112
rmsle                   0.0353167  0.00130899   0.0337463     0.0375869     0.0355834     0.0327149     0.0369519

Scoring History: 


timestamp            duration    number_of_trees    training_rmse       training_mae        training_deviance
---  -------------------  ----------  -----------------  ------------------  ------------------  -------------------
     2019-04-08 17:25:27  16.816 sec  0.0                181.38238043116908  142.3998489274107   32899.56793087735
     2019-04-08 17:25:27  16.882 sec  5.0                107.42392635466818  84.24003602830221   11539.899953453174
     2019-04-08 17:25:27  16.935 sec  10.0               63.894083815772795  50.023308890770906  4082.453946656999
     2019-04-08 17:25:27  16.985 sec  15.0               38.450969966077004  30.034970414165908  1478.4770913321556
     2019-04-08 17:25:27  17.034 sec  20.0               25.653038601608305  19.734860646038076  658.0783894956057
---  ---                  ---         ---                ---                 ---                 ---
     2019-04-08 17:25:30  19.344 sec  270.0              1.7050142079828006  1.1700358492018363  2.9070734494232164
     2019-04-08 17:25:30  19.393 sec  275.0              1.6909674134211323  1.1622558060178851  2.859370793252155
     2019-04-08 17:25:30  19.451 sec  280.0              1.677069230591134   1.1544659421841645  2.812561204195539
     2019-04-08 17:25:30  19.520 sec  285.0              1.66171661485495    1.1441283714813113  2.761302108084994
     2019-04-08 17:25:30  19.534 sec  286.0              1.659672829370749   1.1420248950973013  2.7545139005515074


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
registered  2.56858e+09            1                    0.918241
casual      1.5348e+08             0.0597529            0.0548675
hr          4.37214e+07            0.0170216            0.0156299
instant     9.33502e+06            0.00363431           0.00333717
workingday  7.97255e+06            0.00310387           0.0028501
dteday      5.19893e+06            0.00202404           0.00185856
temp        3.35408e+06            0.00130581           0.00119905
weekday     2.51315e+06            0.000978418          0.000898423
atemp       2.09453e+06            0.000815442          0.000748772
hum         538359                 0.000209594          0.000192457
season      207410                 8.07489e-05          7.41469e-05
weathersit  139702                 5.43889e-05          4.99421e-05
mnth        64992.2                2.53027e-05          2.3234e-05
windspeed   55394.5                2.15662e-05          1.98029e-05
yr          19290.6                7.51022e-06          6.89619e-06
holiday     9624.22                3.7469e-06           3.44055e-06

In [142]:
aml.leader.algo

'gbm'

## Ensemble Exploration

In [143]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
aml_leaderboard_df

model_id  mean_residual_deviance  \
0                        GBM_2_AutoML_20190408_171007               10.398930   
1                        GBM_2_AutoML_20190408_194251               12.792848   
2                        GBM_1_AutoML_20190408_171007               13.248308   
3                        GBM_4_AutoML_20190408_190707               14.211758   
4                        GBM_3_AutoML_20190408_194251               14.299879   
5                    XGBoost_1_AutoML_20190408_171007               14.936004   
6                        GBM_3_AutoML_20190408_190707               15.203899   
7                    XGBoost_3_AutoML_20190408_194251               15.983771   
8                        GBM_1_AutoML_20190408_190707               19.151410   
9                    XGBoost_2_AutoML_20190408_190707               19.551146   
10                   XGBoost_1_AutoML_20190408_194251               19.981404   
11                   XGBoost_1_AutoML_20190408_190707               20.960508   
12                       GBM_2_AutoML_20190408_190707               22.332141   
13                       GBM_3_AutoML_20190408_171007               22.350299   
14   StackedEnsemble_AllModels_AutoML_20190408_171007               23.104649   
15                   XGBoost_3_AutoML_20190408_171007               23.432444   
16                   XGBoost_2_AutoML_20190408_194251               25.207378   
17                       GBM_4_AutoML_20190408_171007               25.881513   
18                       GBM_4_AutoML_20190408_194251               28.378288   
19                   XGBoost_3_AutoML_20190408_190707               28.958293   
20              DeepLearning_1_AutoML_20190408_190707               31.081973   
21                   XGBoost_2_AutoML_20190408_171007               31.390413   
22                       GBM_1_AutoML_20190408_194251               31.636260   
23              DeepLearning_1_AutoML_20190408_194251               34.905251   
24          GLM_grid_1_AutoML_20190408_190707_model_1               38.387049   
25          GLM_grid_1_AutoML_20190408_194251_model_1               38.387049   
26          GLM_grid_1_AutoML_20190408_171007_model_1               38.387049   
27                       XRT_1_AutoML_20190408_190707               43.280700   
28      XGBoost_grid_1_AutoML_20190408_190707_model_1               44.578618   
29                       DRF_1_AutoML_20190408_171007               44.997852   
30                       XRT_1_AutoML_20190408_194251               45.578332   
31                       DRF_1_AutoML_20190408_194251               47.859861   
32              DeepLearning_1_AutoML_20190408_171007               51.970420   
33                       DRF_1_AutoML_20190408_190707               55.530565   
34  StackedEnsemble_BestOfFamily_AutoML_20190408_1...               58.092277   
35                       GBM_5_AutoML_20190408_194251               59.698413   
36      XGBoost_grid_1_AutoML_20190408_194251_model_1               75.332722   
37                       GBM_5_AutoML_20190408_171007               82.943169   
38                       GBM_5_AutoML_20190408_190707               83.974272   
39                       XRT_1_AutoML_20190408_171007              156.118172   
40          GBM_grid_1_AutoML_20190408_194251_model_2              188.252075   
41      XGBoost_grid_1_AutoML_20190408_194251_model_2              453.898132   
42  DeepLearning_grid_1_AutoML_20190408_194251_mod...              625.837691   
43  DeepLearning_grid_1_AutoML_20190408_194251_mod...              854.134260   
44  DeepLearning_grid_1_AutoML_20190408_194251_mod...              974.398082   
45  DeepLearning_grid_1_AutoML_20190408_190707_mod...             1009.056343   
46  DeepLearning_grid_1_AutoML_20190408_194251_mod...             1154.372204   
47  DeepLearning_grid_1_AutoML_20190408_190707_mod...             1703.482385   
48      XGBoost_grid_1_AutoML_20190408_190707_model_2             2898.05391

## Getting Models

### Parameters for GBM_2_AutoML_20190408_171007

In [144]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[0])

In [145]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'GBM_2_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/GBM_2_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'score_tree_interval': {'default': 0, 'actual': 5},
 'fold_assignment': {'default': 'AUTO',

### Parameters for XGBoost_1_AutoML_20190408_171007	

In [146]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[5])

In [147]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'XGBoost_1_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/XGBoost_1_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'fold_assignment': {'default': 'AUTO', 'actual': 'Modulo'},
 'fold_column': {'defau

### Parameters for DRF_1_AutoML_20190408_171007

In [148]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[29])

In [149]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'DRF_1_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/DRF_1_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'score_tree_interval': {'default': 0, 'actual': 0},
 'fold_assignment': {'default': 'AUTO',

## Model with runtime 1850 seconds

In [150]:
# Assume the following are passed by the user from the web interface

'''
Need a user id and project id?

'''
target='cnt' 
data_file='hour.csv'
run_time=1850
run_id='SOME_ID_20180617_221531' # Just some arbitrary ID
server_path='Users/newzysharma/Desktop/Desktop/Machine_Learning/Project'
classification=False
scale=False
max_models=None
balance_y=False # balance_classes=balance_y
balance_threshold=0.2
project ="HyperparameterDB_Project"  # project_name = project

In [151]:
# Use local data file or download from some type of bucket
import os

data_path=os.path.join(server_path,data_file)
data_path

'Users/newzysharma/Desktop/Desktop/Machine_Learning/Project/hour.csv'

In [152]:
# assign target and inputs for logistic regression
y = target
X = [name for name in hp.columns if name != y]
print(y)
print(X)

cnt
['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered']


In [153]:
# impute missing values
_ = hp[reals].impute(method='mean')
_ = hp[ints].impute(method='median')

if scale:
    hp[reals] = hp[reals].scale()
    hp[ints] = hp[ints].scale()

In [154]:
if classification:
    class_percentage = y_balance=df[y].mean()[0]/(df[y].max()-df[y].min())
    if class_percentage < balance_threshold:
        balance_y=True
        

print(run_time)
type(run_time)

1850


int

## Cross-validate rather than take a test training split with 1850 seconds

In [155]:
# automl
# runs for run_time seconds then builds a stacked ensemble

aml = H2OAutoML(max_runtime_secs=run_time,project_name = project) # init automl, run for 1850 seconds
aml.train(x=X,  
           y=y,
           training_frame=hp)

AutoML progress: |████████████████████████████████████████████████████████| 100%


## Leaderboard

In [156]:
# view leaderboard
lb = aml.leaderboard
lb

model_id mean_residual_deviance rmse mse mae rmsle 
 
 
 GBM_2_AutoML_20190408_171007 10.3989 3.22474 10.3989 1.88964 0.0353651 
 GBM_2_AutoML_20190408_194251 12.7928 3.57671 12.7928 2.13437 0.0450689 
 GBM_3_AutoML_20190408_203115 12.9605 3.60007 12.9605 2.11321 0.0405568 
 GBM_1_AutoML_20190408_171007 13.2483 3.63982 13.2483 2.30598 0.0573338 
 GBM_1_AutoML_20190408_203115 13.4612 3.66895 13.4612 2.19063 0.0428744 
 GBM_4_AutoML_20190408_190707 14.2118 3.76985 14.2118 2.15209 0.0367956 
 GBM_3_AutoML_20190408_194251 14.2999 3.78152 14.2999 2.17271 0.0358335 
 XGBoost_1_AutoML_20190408_171007 14.936 3.86471 14.936 2.25298 0.0426858 
 GBM_3_AutoML_20190408_190707 15.2039 3.89922 15.2039 2.27726 0.0438891 
 XGBoost_3_AutoML_20190408_194251 15.9838 3.99797 15.9838 2.48441 nan

In [157]:
aml.leader

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  GBM_2_AutoML_20190408_171007


ModelMetricsRegression: gbm
** Reported on train data. **

MSE: 2.7545139005515074
RMSE: 1.659672829370749
MAE: 1.1420248950973013
RMSLE: 0.026260655890205944
Mean Residual Deviance: 2.7545139005515074

ModelMetricsRegression: gbm
** Reported on cross-validation data. **

MSE: 10.398929973581447
RMSE: 3.2247371944984056
MAE: 1.8896396840192615
RMSLE: 0.03536506290104928
Mean Residual Deviance: 10.398929973581447
Cross-Validation Metrics Summary: 


mean       sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
----------------------  ---------  -----------  ------------  ------------  ------------  ------------  ------------
mae                     1.88964    0.0331531    1.88882       1.84409       1.95555       1.83307       1.92668
mean_residual_deviance  10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
mse                     10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
r2                      0.999684   3.44699e-05  0.999712      0.999737      0.999593      0.999687      0.999689
residual_deviance       10.3989    1.1037       9.48462       8.7057        13.3016       10.3194       10.1832
rmse                    3.21618    0.166        3.07971       2.95054       3.64714       3.21238       3.19112
rmsle                   0.0353167  0.00130899   0.0337463     0.0375869     0.0355834     0.0327149     0.0369519

Scoring History: 


timestamp            duration    number_of_trees    training_rmse       training_mae        training_deviance
---  -------------------  ----------  -----------------  ------------------  ------------------  -------------------
     2019-04-08 17:25:27  16.816 sec  0.0                181.38238043116908  142.3998489274107   32899.56793087735
     2019-04-08 17:25:27  16.882 sec  5.0                107.42392635466818  84.24003602830221   11539.899953453174
     2019-04-08 17:25:27  16.935 sec  10.0               63.894083815772795  50.023308890770906  4082.453946656999
     2019-04-08 17:25:27  16.985 sec  15.0               38.450969966077004  30.034970414165908  1478.4770913321556
     2019-04-08 17:25:27  17.034 sec  20.0               25.653038601608305  19.734860646038076  658.0783894956057
---  ---                  ---         ---                ---                 ---                 ---
     2019-04-08 17:25:30  19.344 sec  270.0              1.7050142079828006  1.1700358492018363  2.9070734494232164
     2019-04-08 17:25:30  19.393 sec  275.0              1.6909674134211323  1.1622558060178851  2.859370793252155
     2019-04-08 17:25:30  19.451 sec  280.0              1.677069230591134   1.1544659421841645  2.812561204195539
     2019-04-08 17:25:30  19.520 sec  285.0              1.66171661485495    1.1441283714813113  2.761302108084994
     2019-04-08 17:25:30  19.534 sec  286.0              1.659672829370749   1.1420248950973013  2.7545139005515074


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
registered  2.56858e+09            1                    0.918241
casual      1.5348e+08             0.0597529            0.0548675
hr          4.37214e+07            0.0170216            0.0156299
instant     9.33502e+06            0.00363431           0.00333717
workingday  7.97255e+06            0.00310387           0.0028501
dteday      5.19893e+06            0.00202404           0.00185856
temp        3.35408e+06            0.00130581           0.00119905
weekday     2.51315e+06            0.000978418          0.000898423
atemp       2.09453e+06            0.000815442          0.000748772
hum         538359                 0.000209594          0.000192457
season      207410                 8.07489e-05          7.41469e-05
weathersit  139702                 5.43889e-05          4.99421e-05
mnth        64992.2                2.53027e-05          2.3234e-05
windspeed   55394.5                2.15662e-05          1.98029e-05
yr          19290.6                7.51022e-06          6.89619e-06
holiday     9624.22                3.7469e-06           3.44055e-06

In [158]:
aml.leader.algo

'gbm'

## Ensemble Exploration

In [159]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
aml_leaderboard_df

model_id  mean_residual_deviance  \
0                        GBM_2_AutoML_20190408_171007               10.398930   
1                        GBM_2_AutoML_20190408_194251               12.792848   
2                        GBM_3_AutoML_20190408_203115               12.960531   
3                        GBM_1_AutoML_20190408_171007               13.248308   
4                        GBM_1_AutoML_20190408_203115               13.461223   
5                        GBM_4_AutoML_20190408_190707               14.211758   
6                        GBM_3_AutoML_20190408_194251               14.299879   
7                    XGBoost_1_AutoML_20190408_171007               14.936004   
8                        GBM_3_AutoML_20190408_190707               15.203899   
9                    XGBoost_3_AutoML_20190408_194251               15.983771   
10                   XGBoost_3_AutoML_20190408_203115               16.064945   
11              DeepLearning_1_AutoML_20190408_203115               17.888279   
12                       GBM_1_AutoML_20190408_190707               19.151410   
13                       GBM_2_AutoML_20190408_203115               19.513417   
14                   XGBoost_2_AutoML_20190408_190707               19.551146   
15                   XGBoost_1_AutoML_20190408_194251               19.981404   
16                   XGBoost_1_AutoML_20190408_203115               20.813417   
17                   XGBoost_1_AutoML_20190408_190707               20.960508   
18                       GBM_2_AutoML_20190408_190707               22.332141   
19                       GBM_3_AutoML_20190408_171007               22.350299   
20   StackedEnsemble_AllModels_AutoML_20190408_171007               23.104649   
21                   XGBoost_3_AutoML_20190408_171007               23.432444   
22                   XGBoost_2_AutoML_20190408_203115               24.924710   
23                   XGBoost_2_AutoML_20190408_194251               25.207378   
24                       GBM_4_AutoML_20190408_171007               25.881513   
25                       GBM_4_AutoML_20190408_194251               28.378288   
26                   XGBoost_3_AutoML_20190408_190707               28.958293   
27                       GBM_5_AutoML_20190408_203115               29.071952   
28              DeepLearning_1_AutoML_20190408_190707               31.081973   
29                   XGBoost_2_AutoML_20190408_171007               31.390413   
..                                                ...                     ...   
55          GBM_grid_1_AutoML_20190408_194251_model_2              188.252075   
56      XGBoost_grid_1_AutoML_20190408_203115_model_2              235.889029   
57          GBM_grid_1_AutoML_20190408_203115_model_4              246.075808   
58          GBM_grid_1_AutoML_20190408_203115_model_2              280.861071   
59  DeepLearning_grid_1_AutoML_20190408_203115_mod...              360.272455   
60      XGBoost_grid_1_AutoML_20190408_194251_model_2              453.898132   
61  DeepLearning_grid_1_AutoML_20190408_203115_mod...              471.079686   
62          GBM_grid_1_AutoML_20190408_203115_model_6              497.932351   
63  DeepLearning_grid_1_AutoML_20190408_194251_mod...              625.837691   
64  DeepLearning_grid_1_AutoML_20190408_194251_mod...              854.134260   
65  DeepLearning_grid_1_AutoML_20190408_194251_mod...              974.398082   
66          GBM_grid_1_AutoML_20190408_203115_model_7              985.511429   
67  DeepLearning_grid_1_AutoML_20190408_190707_mod...             1009.056343   
68  DeepLearning_grid_1_AutoML_20190408_203115_mod...             1009.740664   
69  DeepLearning_grid_1_AutoML_20190408_194251_mod...             1154.372204   
70  DeepLearning_grid_1_AutoML_20190408_190707_mod...             1703.482385   
71  DeepLearning_grid_1_AutoML_20190408_203115_mod...             1770.021831   
72      XGBoost_grid_1_AutoML_20190408_190707_model_2             2898.05391

## Getting Models

### Parameters for GBM_2_AutoML_20190408_171007

In [160]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[0])

In [161]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'GBM_2_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/GBM_2_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'score_tree_interval': {'default': 0, 'actual': 5},
 'fold_assignment': {'default': 'AUTO',

### Parameters for XGBoost_1_AutoML_20190408_190707

In [162]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[17])

In [163]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'XGBoost_1_AutoML_20190408_190707',
   'type': 'Key<Model>',
   'URL': '/3/Models/XGBoost_1_AutoML_20190408_190707'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'fold_assignment': {'default': 'AUTO', 'actual': 'Modulo'},
 'fold_column': {'defau

### Parameters for StackedEnsemble_AllModels_AutoML_20190408_171007

In [164]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[20])

In [165]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'StackedEnsemble_AllModels_AutoML_20190408_171007',
   'type': 'Key<Model>',
   'URL': '/3/Models/StackedEnsemble_AllModels_AutoML_20190408_171007'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour.hex'}},
 'response_column': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ColSpecifierV3',
    'schema_type': 'VecSpecifier'},
   'column_name': 'cnt',
   'is_member_of_frames': None}},
 'validation_frame': {'default': None, 'actual': None},
 'blending_frame': {'default': None, 'actual': None},
 'base_models': {'default': [],
  'actual': [{'__meta': {'schema_version': 3,
     'schema_name': 'ModelKey

# CONCLUSION

<table style="width:50%">
  <tr>
      <th>Runtime of model in Number of Seconds<br></th>
    <th>Models Generated</th> 
  </tr>
    
   <tr>
    <td>500</td>
    <td>7</td> 
  </tr>
    
  <tr>
    <td>1000</td>
    <td>13</td> 
  </tr>
  
  <tr>
    <td>1350</td>
    <td>33</td> 
  </tr>
  
   <tr>
    <td>1500</td>
    <td>56</td> 
  </tr>
  
   <tr>
    <td>1850</td>
    <td>84<td> 
  </tr>
    
</table>

# CONTRIBUTION

# CITATIONS

https://github.com/nikbearbrown/CSYE_7245/blob/master/H2O/H2O_automl_model.ipynb

# LICENSE


Copyright 2019 Newzy Sharma 

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
documentation files (the "Software"), to deal in the Software without restriction, including without limitation the 
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the 
Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE 
WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.